lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,402 @@
+    """
+    try:
+        ProviderSchema(app_data=FivegCoreGnbProviderAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class GnbAvailableEvent(EventBase):
+    """Dataclass for the `fiveg_core_gnb` request event."""
+
+    def __init__(self, handle: Handle, relation_id: int):
+        """Set relation id.
+
+        Args:
+            handle (Handle): Juju framework handle.
+            relation_id : ID of the relation.
+        """
+        super().__init__(handle)
+        self.relation_id = relation_id
+
+    def snapshot(self) -> dict:
+        """Return event data.
+
+        Returns:
+            (dict): contains the relation ID.
+        """
+        return {
+            "relation_id": self.relation_id,
+        }
+
+    def restore(self, snapshot: dict) -> None:
+        """Restore event data.
+
+        Args:
+            snapshot (dict): contains the relation ID.
+        """
+        self.relation_id = snapshot["relation_id"]
+
+
+class FivegCoreGnbProviderCharmEvents(CharmEvents):
+    """Custom events for the FivegCoreGnbProvider."""
+
+    gnb_available = EventSource(GnbAvailableEvent)
+
+
+class FivegCoreGnbProvides(Object):
+    """Class to be instantiated by provider of the `fiveg_core_gnb`."""
+
+    on = FivegCoreGnbProviderCharmEvents()  # type: ignore
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Observe relation joined event.
+
+        Args:
+            charm: Juju charm
+            relation_name (str): Relation name
+        """
+        self.relation_name = relation_name
+        self.charm = charm
+        super().__init__(charm, relation_name)
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+
+    def publish_fiveg_core_gnb_information(
+        self, relation_id: int, tac: int, plmns: list[PLMNConfig]
+    ) -> None:
+        """Set TAC and PLMNs in the relation data.
+
+        Args:
+            relation_id (str): Relation ID.
+            tac (int): Tracking Area Code.
+            plmns (list[PLMNConfig]): Configured PLMNs.
+        """
+        if not data_matches_provider_schema(
+            data={"tac": tac, "plmns": plmns}
+        ):
+            raise ValueError(f"Invalid fiveG core gNB data: {tac}, {plmns}")
+        relation = self.model.get_relation(
+            relation_name=self.relation_name, relation_id=relation_id
+        )
+        if not relation:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+        relation.data[self.charm.app].update({"tac": str(tac), "plmns": json.dumps(plmns)})
+
+    def _on_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Triggered whenever a requirer charm joins the relation.
+
+        Args:
+            event (RelationJoinedEvent): Juju event
+        """
+        self.on.gnb_available.emit(relation_id=event.relation.id)
```

Comment:
We should be firing this event only if the gNB name is present in the relation data, shouldn't we?